### ISSUE TITLE
can't knit index.Rmd

### ISSUE THREAD

**Issue by ricardd** [pbs-assess/csasdowntex#19]
Not expecting an answer, just opening this issue to document the ongoing problem I have when trying to knit index.Rmd.

```
"C:/Program Files/RStudio/bin/pandoc/pandoc" +RTS -K512m -RTS resdoc.utf8.md --to latex --from markdown+autolink_bare_uris+ascii_identifiers+tex_math_single_backslash --output resdoc.tex --table-of-contents --toc-depth 3 --template "C:\~\R\win-library\3.5\csasdown\csas-tex\res-doc.tex" --number-sections --highlight-style tango --latex-engine pdflatex --top-level-division=chapter --wrap=none --variable tables=yes --standalone --filter pandoc-citeproc 
output file: resdoc.knit.md

! Missing \endcsname inserted.
<to be read again> 
                   \protect 
l.15 

Error: Failed to compile resdoc.tex. See resdoc.log for more info.
Execution halted
```
and the tail of the log file:
```
! Missing \endcsname inserted.
<to be read again> 
                   \protect 
l.15 
```
It was initially suggested that MiKTeX was to blame, but I am getting the same error after uninstalling MiKTeX and installing tinytex from R instead (which basically installs TeXLive).

**Comment by seananderson** [comment]
@brookemdavis is having the same problem... which isn’t very helpful unless on the off chance we can figure out what’s in common between your setups.

I don’t know the root of the problem (we were guessing at character encoding problems for Dan, but don’t really know), but it almost certainly comes down to an offending package that’s getting loaded. 

It’s hard for me to debug without a computer that gets this error, but a debugging strategy would be to take the _book/resdoc.tex file, put it back in the folder up a level, and try stripping out packages and code until this problem disappears when running pdflatex on it. Much of it is in the style file, so you’d need to change the path to the .sty file to be local to that folder (in the tex) and edit the .sty as needed.

**Comment by brookemdavis** [comment]
I tried to run it on another computer and have the same issue.

**Comment by brookemdavis** [comment]
Where do I find the .sty file? I'm not seeing it anywhere.

**Comment by ricardd** [comment]
@brookemdavis, the style files are in the inst folder 

csasdown /inst/csas-style

**Comment by brookemdavis** [comment]
Sorry I think I might be in over my head with this, so I'm not great at debugging. 

I put res-doc.sty in the same folder that the package pasted everything to. So in that folder right now I have folders: bib, csl, data, figure, knitr-cache-tex, knitr-figs-pdf, templates. And files: _bookdown.ylm, 01-Chap1.Rmd .... 99-references.RMD, index.Rmd, resdoc.log, resdoc.Rmd, res-doc.sty, resdoc.tex

I opened up resdoc.tex in Rstudio and tried to "Compile PDF" but it can't find res-doc.sty

So I figure the error is from the .tex file directing it to my R package library. So I comment out line 11 (\newcommand{\locRepo}{C:/~/R/win-library/3.5/csasdown/csas-style})  and change line 14 from: \usepackage{\locRepo/res-doc}  to \usepackage{res-doc}

It looks like that fixed the .sty file directory issue...

Then my new error is LaTeX Error: File 'nccmath.sty' not found.

So I guess this is the package that is missing?

**Comment by brookemdavis** [comment]
Not sure if I'm looking in the right place but if (in R console) I do:
library(tinytex)
tl_pkgs()

to see which tex packages I have -- I don't see "nccmath"

**Comment by brookemdavis** [comment]
I've done a but more (somewhat blind) messing around and it seems like a whole bunch of packages are missing for some reason. 

So after I got the nccmath error I tried to install it using tinytex commands
tlmgr_install(pkgs = "nccmath")

It couldn't install, so after a quick google, I saw it's in a package called "ncctools", so I ran
tlmgr_install(pkgs = "ncctools")
Which seemed to work and installed that package.

Then tried to re-compile the pdf. Same error, different package. So I installed that package, or one that contains it. I did this a number of times downloading packages lineno, fancyhdr, psnfss (error thrown for helvet.sty), tocloft, textpos, and finally got caught up on "everyshi" which seems to be called in from textpos. Can't figure out how to get this one.

Why would I be missing so many packages?

**Comment by brookemdavis** [comment]
Sorry I know I've probably been extremely annoying with this, and am blowing up your notifications. 

I was able to download all the needed packages by using the tinytex command

latexmk("resdoc.tex", install_packages=T)

rather than having to do them one at a time.

So I thought I was rolling after that, but I still have a load of errors, and fail to knit index.rmd or compile resdoc.tex.

Not sure if this will work but here's a snapshot of some of the errors from trying to compile resdoc.tex:

(https://user-images.githubusercontent.com/7797141/49906339-4edb3000-fe26-11e8-9b8f-7b9b8a2d3d44.png)

**Comment by seananderson** [comment]
Hmm, you might have a different problem then, or maybe a combination of problems?

It sounds like you're using tinytex. In that case, are you compiling the .tex file with `tinytex::pdflatex('resdoc.tex')`?

Ordinarily, when `bookdown::render_book()` is run (e.g. via the knit button) or `rmarkdown::render()` is run, the necessary TeX packages should be installed.

I see your latest message now...

**Comment by seananderson** [comment]
Your latest errors are coming from the image files not being in the right place. Ordinarily you wouldn't have to worry about this, but because you are trying to compile it as a separate .tex file you do have to.

You need to make sure the .png files are in the right place in the `res-doc` folder that has the .sty file. Make sure all the contents of the folder are in there:
https://github.com/pbs-assess/csasdown/tree/master/inst/csas-style

I'm assuming you're doing this on the default csasdown draft document that it creates and that you had installed or updated csasdown immediately before starting.

**Comment by brookemdavis** [comment]
Yes I am just trying to knit the default draft document to use it as an example. And yes I was just following along with the instructions in the csasdown readme to install and knit the draft document.

I got the directories for the images and style files fixed. But still not running.

![errors](https://user-images.githubusercontent.com/7797141/49958521-f0fb2680-febf-11e8-9705-eff803cdcb32.png)

**Comment by seananderson** [comment]
I think we're back to your original problem now.

Here's a folder with a working example of the .tex file:
https://www.dropbox.com/sh/8kckkk2k7kuuvmd/AAD5eHPNSU7olnZ9JCnxZ7Efa?dl=1

You should be able to create it with:
```r
tinytex::latexmk("resdoc.tex")
```

assuming you start in the working directory with the TeX file.

Or, once that's been run once and all the TeX packages are installed by tinytex, you should also be able to compile it on any command line with `pdflatex resdoc.tex`. If you are still getting that error about `Missing \endcsname inserted.`, then that's when we need to start troubleshooting by pulling out TeX packages from the style file and cutting out code until we get things to work to figure out what the problem is.

**Comment by brookemdavis** [comment]
Still can't compile and get similar errors.

![errors](https://user-images.githubusercontent.com/7797141/49959163-a5e21300-fec1-11e8-856c-7ee12f380694.png)

**Comment by brookemdavis** [comment]
I can't narrow it down to an exact package, but the main problems are definitely in \frontmatter, and making "longtable"'s.
The pdf will compile with line 111 (\frontmatter) and lines 257-279 and 320-378 (two places "longtable"'s are created) commented out.

**Comment by seananderson** [comment]
Interesting. Does it work for you if you just comment out line 111 or just comment out the longtable section or does it require a combination of the two?

Leaving the longtable part out (i.e. leaving it commented out), and assuming the frontmatter on its own is a problem, can you try commenting out sections of frontmatter in `csas-style/res-doc.sty` to figure out which are the problem lines? See the section starting on line 604 of  `csas-style/res-doc.sty`.

**Comment by brookemdavis** [comment]
I have to have them both commented out to make it run. Will look into frontmatter and get back to you.

**Comment by seananderson** [comment]
Hmm. Weird. My first guest might be the TableOfContents at the bottom or the abstract or the `@` symbol in the csas e-mail address.

**Comment by brookemdavis** [comment]
OK I feel like I'm a crazy person, but it's working with frontmatter now. Just the longtables that seem to be an issue.

**Comment by seananderson** [comment]
OK, well that narrows it down.

Can you compile this document?

```tex
\documentclass[11pt]{book}

\usepackage{longtable,array}
\usepackage{booktabs}

\begin{document}

\begin{longtable}[]{@{}lr@{}}
\caption{\label{tab:maxdelays}Maximum Delays by Airline}\tabularnewline
\toprule
Airline & Max Arrival Delay\tabularnewline
\midrule
\endfirsthead
\toprule
Airline & Max Arrival Delay\tabularnewline
\midrule
\endhead
Alaska Airlines Inc. & 104\tabularnewline
American Airlines Inc. & 119\tabularnewline
Delta Air Lines Inc. & 119\tabularnewline
Frontier Airlines Inc. & 128\tabularnewline
Hawaiian Airlines Inc. & 7\tabularnewline
JetBlue Airways & 15\tabularnewline
SkyWest Airlines Inc. & 81\tabularnewline
Southwest Airlines Co. & 178\tabularnewline
United Air Lines Inc. & 168\tabularnewline
US Airways Inc. & 19\tabularnewline
Virgin America & 66\tabularnewline
\bottomrule
\end{longtable}

\end{document}

```

If not, can you try playing around with the alignment, the caption, and the booktabs?

**Comment by brookemdavis** [comment]
That does compile! Weird.

**Comment by seananderson** [comment]
And I assume it also works if you reference it? `Table~\ref{tab:maxdelays}.`

Could you take `resdoc.tex` and `res-doc.sty` and try removing code until you get a minimal example that breaks if you uncomment the longtable code and works if you leave it in? I assume there is some conflict with something else in the document.

Alternatively, take that simple example from above and start adding the pre-amble elements from res-doc.sty until it breaks.

**Comment by ricardd** [comment]
In a slightly twisted way, I am happy that @brookemdavis is experiencing the same problem as me. 

I can also compile the MWE above with the delays by airlines, including with the table is referenced in the document.

And BTW, I get the same error when trying to compile a PDF with the Science Response.

@seananderson, can we agree on a troubleshooting plan to finally identify what is causing the problem here? I find it daunting right now because I don't know where to start.

**Comment by seananderson** [comment]
My comments above are my best guess. Start by making sure that your resdoc.tex file works if you comment out the 2 longtable sections (download the .zip file from above) like @brookemdavis did above to make sure you're having the same problem. From my googling it seems like that error is very generic. If that works then I guess a short-term solution is to just not use longtables, which I think we turn on by default via an argument with csas_table(), but isn't always essential. You could try removing the tables from the .Rmd and recompiling to see if that works.

Then either start adding content from the broken .tex and .sty file to the MWE longtable above until it breaks or try removing code from the resdoc.tex and .sty until they stop breaking to isolate the problem.

The first thing I would try would be adding all the packages from the .sty file into the preamble of the MWE (then try running it) followed by copying the entire preamble from the .sty file into the MWE since the most likely thing seems to be two packages causing a conflict somehow. Note that when the style file is loaded it's effectively the same as dumping all that code into the .tex file.

**Comment by ricardd** [comment]
Resolved for me.

The error I was getting was seemingly caused by a permission error where the LaTeX compilation failed because the style file could not be read. 

Compilation from RStudio using the "Knit" button resulted in the following error:

```
output file: resdoc.knit.md

! Missing \endcsname inserted.
<to be read again> 
                   \protect 
l.15 

Error: Failed to compile resdoc.tex. See resdoc.log for more info.
Execution halted
```

Line 11 of the "resdoc.tex" generated was modified from

```
\newcommand{\locRepo}{C:/~/R/win-library/3.5/csasdown/csas-style}
```

to

```
\newcommand{\locRepo}{.}
```

copying the LaTeX style file and the images/ directory in the same place as index.Rmd, and the PDF compiled correctly. I had to make sure to delete the resdoc.aux prior to running pdflatex again.

Thanks @seananderson for the help. 

Not sure whether locally copying the files read from the R package (the LaTeX style file and the different bitmap images) would be an appropriate solution here.
